src/tests/frontend/specs/easysync-assembler.js

```diff
@@ -18,6 +19,171 @@ describe('easysync-assembler', function () {
+    const x = '-c*3*4+6*3*4|1+1*3*4|9+f*3*4+k=5';
+    const iter = Changeset.opIterator(x);
+    const assem = Changeset.smartOpAssembler();
+    while (iter.hasNext()) assem.append(iter.next());
+    assem.endDocument();
+    expect(assem.toString()).to.equal('-c*3*4|a+m*3*4+k');
+  });
+
+  it('smartOpAssembler merge consecutive - ops without multiline', async function () {
+    const x = '-c-6-1-9=5';
+    const iter = Changeset.opIterator(x);
+    const assem = Changeset.smartOpAssembler();
+    while (iter.hasNext()) assem.append(iter.next());
+    assem.endDocument();
+    expect(assem.toString()).to.equal('-s');
+  });
+
+  it('smartOpAssembler merge consecutive - ops with multiline', async function () {
+    const x = '-c-6|1-1|9-f-k=5';
+    const iter = Changeset.opIterator(x);
+    const assem = Changeset.smartOpAssembler();
+    while (iter.hasNext()) assem.append(iter.next());
+    assem.endDocument();
+    expect(assem.toString()).to.equal('|a-y-k');
+  });
+
+  it('smartOpAssembler merge consecutive = ops without multiline', async function () {
+    const x = '-c*3*4=6*2*4=1*3*4=f*3*4=2*3*4=a=k=5';
+    const iter = Changeset.opIterator(x);
+    const assem = Changeset.smartOpAssembler();
+    while (iter.hasNext()) assem.append(iter.next());
+    assem.endDocument();
+    expect(assem.toString()).to.equal('-c*3*4=6*2*4=1*3*4=r');
+  });
+
+  it('smartOpAssembler merge consecutive = ops with multiline', async function () {
+    const x = '-c*3*4=6*2*4|1=1*3*4|9=f*3*4|2=2*3*4=a*3*4=1=k=5';
+    const iter = Changeset.opIterator(x);
+    const assem = Changeset.smartOpAssembler();
+    while (iter.hasNext()) assem.append(iter.next());
+    assem.endDocument();
+    expect(assem.toString()).to.equal('-c*3*4=6*2*4|1=1*3*4|b=h*3*4=b');
+  });
+
+  it('smartOpAssembler ignore + ops with ops.chars === 0', async function () {
+    const x = '-c*3*4+6*3*4+0*3*4+1+0*3*4+1';
+    const iter = Changeset.opIterator(x);
+    const assem = Changeset.smartOpAssembler();
+    while (iter.hasNext()) assem.append(iter.next());
+    assem.endDocument();
+    expect(assem.toString()).to.equal('-c*3*4+8');
+  });
+
+  it('smartOpAssembler ignore - ops with ops.chars === 0', async function () {
+    const x = '-c-6-0-1-0-1';
+    const iter = Changeset.opIterator(x);
+    const assem = Changeset.smartOpAssembler();
+    while (iter.hasNext()) assem.append(iter.next());
+    assem.endDocument();
+    expect(assem.toString()).to.equal('-k');
+  });
+
+  it('smartOpAssembler append - op with text', async function () {
+    const assem = Changeset.smartOpAssembler();
+    const pool = poolOrArray([
+      'attr1,1',
+      'attr2,2',
+      'attr3,3',
+      'attr4,4',
+      'attr5,5',
+    ]);
+
+    assem.appendOpWithText('-', 'test', '*3*4*5', pool);
+    assem.appendOpWithText('-', 'test', '*3*4*5', pool);
+    assem.appendOpWithText('-', 'test', '*1*4*5', pool);
+    assem.endDocument();
+    expect(assem.toString()).to.equal('*3*4*5-8*1*4*5-4');
```

Comment:
Thanks. I'll move this to another issue/PR so we can isolate it from the other changes. Not sure if it's explicitly forbidden (ie if it can cause any harm), but I can't imagine any real world scenario where remove ops will contain attributes.